consensus/ballot_state.go

```diff
@@ -529,21 +531,59 @@ func (s *BallotState) RelevantRange(x SlotValue) (int, int) {
 	min, max := 0, 0
 	for _, message := range s.M {
 		a, b := message.RelevantRange(x)
-		min, max = RangeUnion(min, max, a, b) 
+		min, max = RangeUnion(min, max, a, b)
 	}
 	return min, max
 }
 
+// Returns the max ballot number that a blocking set of nodes are talking about.
+func (s *BallotState) MaxActionableBallotNumber() int {
+	numberToNodes := make(map[int][]string)
+
+	for node, message := range s.M {
+		maxN := message.MaxN()
+		numberToNodes[maxN] = append(numberToNodes[maxN], node)
+	}
+
+	i := 0
+	nKeys := make([]int, len(numberToNodes))
+	for n := range numberToNodes {
+		nKeys[i] = n
+		i++
+	}
+
+	sort.Sort(sort.Reverse(sort.IntSlice(nKeys)))
+
+	nodesAbove := []string{}
+	max := 0
+
+	for _, n := range nKeys {
+		nodesAbove = append(nodesAbove, numberToNodes[n]...)
+		if (s.D.BlockedBy(nodesAbove)) {
+			max = n
```

Comment:
you don't really need max, you can just return n here and return 0 if the for loop terminates

Reply:
Good point. Will fix.